legacy/messaging/src/main/java/org/jboss/as/messaging/MigrateOperation.java

```diff
@@ -0,0 +1,370 @@
+ *     This returns a list of :add operations for each messaging resources.</li>
+ *     <li>:add the new org.widlfy.extension.messaging-activemq extension</li>
+ *     <li>for each messaging resources, transform the :add operations to add the
+ *     corresponding resource to the new messaging-activemq subsystem.
+ *     In this step, changes to the resources model are taken into account
+ *     (e.g. cluster-connection's connector-ref is named connector-name in the new messaging-activemq subsystem.)</li>
+ *     <li>:remove the messaging subsystem</li>
+ * </ul>
+ *
+ * The companion <code>:describe-migration</code> operation will return a list of all the actual operations that would be
+ * performed during the invocation of the <code>:migrate</code> operation.
+ *
+ * @author <a href="http://jmesnil.net/">Jeff Mesnil</a> (c) 2015 Red Hat inc.
+ */
+
+public class MigrateOperation implements OperationStepHandler {
+
+    private static final String MESSAGING_ACTIVEMQ_EXTENSION = "org.wildfly.extension.messaging-activemq";
+    private static final String MESSAGING_ACTIVEMQ_MODULE = "org.wildfly.extension.messaging-activemq";
+
+    private static final String NEW_ENTRY_SUFFIX = "-new";
+
+    private static final OperationStepHandler DESCRIBE_MIGRATION_INSTANCE = new MigrateOperation(true);
+    private static final OperationStepHandler MIGRATE_INSTANCE = new MigrateOperation(false);
+
+    private static final AttributeDefinition ADD_LEGACY_ENTRIES = SimpleAttributeDefinitionBuilder.create("add-legacy-entries", BOOLEAN)
+            .setDefaultValue(new ModelNode(false))
+            .build();
+    private static final AttributeDefinition DRY_RUN = SimpleAttributeDefinitionBuilder.create("dry-run", BOOLEAN)
+            .setDefaultValue(new ModelNode(true))
+            .build();
+
+    private final boolean describe;
+
+    private MigrateOperation(boolean describe) {
+
+        this.describe = describe;
+    }
+
+    static void registerOperations(ManagementResourceRegistration registry, ResourceDescriptionResolver resourceDescriptionResolver) {
+        registry.registerOperationHandler(new SimpleOperationDefinitionBuilder("migrate", resourceDescriptionResolver)
+                        .setParameters(ADD_LEGACY_ENTRIES, DRY_RUN)
+                        .setRuntimeOnly()
+                        .setReplyValueType(ModelType.BOOLEAN)
+                        .setAccessConstraints(SensitiveTargetAccessConstraintDefinition.READ_WHOLE_CONFIG)
+                        .build(),
+                MigrateOperation.MIGRATE_INSTANCE);
+        registry.registerOperationHandler(new SimpleOperationDefinitionBuilder("describe-migration", resourceDescriptionResolver)
+                        .addParameter(ADD_LEGACY_ENTRIES)
+                        .setRuntimeOnly()
+                        .setAccessConstraints(SensitiveTargetAccessConstraintDefinition.READ_WHOLE_CONFIG)
```

Comment:
This could have .setReplyType(ModelType.LIST).setReplyValueType(ModelType.OBJECT). We can't properly describe the object's though, as the params for each op vary. But that's ok.